### PR TITLE
Exit with non-zero status code if allow_other is used but not enabled in fuse config

### PR DESCRIPTION
**Pull request by vibhansa-msft**
Fix for #1081

### CODE REVIEW

File: component/libfuse/libfuse.go [hunk 1]
```diff
@@ -204,6 +206,20 @@ func (lf *Libfuse) Validate(opt *LibfuseOptions) error {
...
 
+	if lf.allowOther {
+		// Validate allow_other is enabled in fuse.conf file or not
+		cmd := exec.Command("bash", "-c", "cat /etc/fuse.conf | grep -i user_allow_other$")
+
+		var errb bytes.Buffer
+		cmd.Stderr = &errb
+		cliOut, err := cmd.Output()
+		data := string(cliOut)
+		if err == nil && data[0] == '#' {
```
Comment (souravgupta-msft) [inline component/libfuse/libfuse.go]: Add case for when `user_allow_other` is not present in /etc/fuse.conf

Comment (vibhansa-msft) [reply component/libfuse/libfuse.go]: done

**Review by souravgupta-msft — APPROVED** [review]

File: component/libfuse/libfuse.go [hunk 2]
```diff
@@ -204,6 +207,23 @@ func (lf *Libfuse) Validate(opt *LibfuseOptions) error {
 		lf.negativeTimeout = defaultNegativeEntryExpiration
 	}
 
+	if lf.allowOther {
+		// Validate allow_other is enabled in fuse.conf file or not
+		cmd := exec.Command("bash", "-c", "cat /etc/fuse.conf | grep -i user_allow_other$")
```
Comment (andyzhangx) [inline component/libfuse/libfuse.go:212]: default value is disabled now, so using `allow_other` option would always return error now?
```
# cat /etc/fuse.conf
# /etc/fuse.conf - Configuration file for Filesystem in Userspace (FUSE)

# Set the maximum number of FUSE mounts allowed to non-root users.
# The default is 1000.
#mount_max = 1000

# Allow non-root users to specify the allow_other or allow_root mount options.
#user_allow_other
```

Comment (andyzhangx) [reply component/libfuse/libfuse.go:212]: cc @cvvz

**Review by gapra-msft — APPROVED** [review]

File: component/libfuse/libfuse.go (hunk 2 above)
Comment (vibhansa-msft) [reply component/libfuse/libfuse.go:212]: even today its failing, it just fails silently. With this change it will start returning non-zero error code.

Comment (andyzhangx) [reply component/libfuse/libfuse.go:212]: the above is the default setting in AKS node, and user could set `-o allow_other` mount option, and it won't return error on AKS node today

Comment (cvvz) [reply component/libfuse/libfuse.go:212]: `-o allow_other` only failed when mount by non-root user. Root user can mount successfully even without configuring `/etc/fuse.conf`.

**Comment by cvvz** [comment]
Please take a look at https://github.com/Azure/azure-storage-fuse/pull/1088. I think it should be the right way to solve this kind of problem.

File: component/libfuse/libfuse.go (hunk 2 above)
Comment (vibhansa-msft) [reply component/libfuse/libfuse.go:212]: If user_allow_other is not enabled in /etc/fuse.conf then it shall not work irrespective of you execute mount with sudo or non-sudo or root user.
In current implementation it fails without giving any error and with this change we are pre-validating the file and returning error if mount is suppose to fail later due to this check.
One question here for the AKS team, if we always mount with allow_other option in AKS why dont we enable user_allow_other in /etc/fuse.conf by default when pod is deployed ?

Comment (vibhansa-msft) [reply component/libfuse/libfuse.go:212]: Ok I verified this and appears root user can mount with allow_other even then /etc/fuse.conf does not have this option set. So I can add a check for uid != 0 while checking the things.

Comment (vibhansa-msft) [reply component/libfuse/libfuse.go:214]: Corrected code to skip this check for root user

**Comment by vibhansa-msft** [comment]
> Please take a look at #1088. I think it should be the right way to solve this kind of problem, instead of handling the `-o allow_other` option alone.

That other PR will not solve this issue as post those changes blobfuse binary will exit with non zero status code but still does not give any hint on what went wrong.

**Comment by cvvz** [comment]
Since this PR has been merged and I'm not the expert of blobfuse V2, so I'm not going to oppose to this pr. 

However, I have to say that with #1088, we can sure get the error code and error message, which is **directly returned from libfuse**.
image

The other concern about this PR is that: if we decide to upgrade libfuse someday and some behavior changed, or if there are some differences from fuse v2 and fuse v3, we may still encounter some problem and have to add more similar changes in blobfuse2 validation logic.

**Comment by andyzhangx** [comment]
@vibhansa-msft pls hold on releasing any new blobfuse2 version, we are setting up a daily e2e test validation on blobfuse2 main branch, to make sure there is no regression after integration with blob csi driver.

**Comment by cvvz** [comment]
@vibhansa-msft Can you test with https://github.com/Azure/azure-storage-fuse/pull/1088 again to see if it can solve this issue? I've tested and it can provide error message from libfuse. Please rethink about this PR. Considering the risk I just raised, I'd prefer we solve this problem by   https://github.com/Azure/azure-storage-fuse/pull/1088, rather than this PR. Thanks

**Pull request merged**